pkg/util/winutil/iisconfig/tagtree.go

```diff
@@ -0,0 +1,157 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+//go:build windows
+
+package iisconfig
+
+import (
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"golang.org/x/sys/windows/registry"
+)
+
+type pathTreeEntry struct {
+	nodes     map[string]*pathTreeEntry
+	ddjson    APMTags
+	appconfig APMTags
+}
+
+func splitPaths(path string) []string {
+	path = filepath.Clean(path)
+	if path == "/" || path == "\\" {
+		return make([]string, 0)
+	}
+	a, b := filepath.Split(path)
+	s := splitPaths(a)
+	return append(s, b)
+}
+
+func findInPathTree(pathtrees map[uint32]*pathTreeEntry, id uint32, urlpath string) (APMTags, APMTags) {
+	// urlpath will come in as something like
+	// /path/to/app
+	// need to build the tree all the way down
+
+	// break down the path
+	pathparts := splitPaths(urlpath)
+
+	if _, ok := pathtrees[id]; !ok {
+		return APMTags{}, APMTags{}
+	}
+	if len(pathparts) == 0 {
+		return pathtrees[id].ddjson, pathtrees[id].appconfig
+	}
+
+	currNode := pathtrees[id]
+
+	for _, part := range pathparts {
+		if _, ok := currNode.nodes[part]; !ok {
+			return currNode.ddjson, currNode.appconfig
+		}
+		currNode = currNode.nodes[part]
+	}
+	return currNode.ddjson, currNode.appconfig
+}
+
+func addToPathTree(pathtrees map[uint32]*pathTreeEntry, siteID string, urlpath string, ddjson, appconfig APMTags) {
+
+	intid, err := strconv.Atoi(siteID)
+	if err != nil {
+		return
+	}
+	id := uint32(intid)
+	// urlpath will come in as something like
+	// /path/to/app
+	// need to build the tree all the way down
+
+	// break down the path
+	pathparts := splitPaths(urlpath)
+
+	if _, ok := pathtrees[id]; !ok {
+		pathtrees[id] = &pathTreeEntry{
+			nodes: make(map[string]*pathTreeEntry),
+		}
+	}
+	if len(pathparts) == 0 {
+		pathtrees[id].ddjson = ddjson
+		pathtrees[id].appconfig = appconfig
+		return
+	}
+
+	currNode := pathtrees[id]
+
+	for _, part := range pathparts {
+		if _, ok := currNode.nodes[part]; !ok {
+			currNode.nodes[part] = &pathTreeEntry{
+				nodes: make(map[string]*pathTreeEntry),
+			}
+		}
+		currNode = currNode.nodes[part]
+	}
+	currNode.ddjson = ddjson
+	currNode.appconfig = appconfig
+	return
```

Comment:
I wonder why would you need to split on pars if whole path can be used as a key?

Reply:
What I've done is made a search tree.
As is demonstrated in the test
/path/to/something is inside of /path.  It's inside of /path/to  but it's not in /path/to/else

if you just search on entire path, you'll get false matches.